File: algorithm/challenges/02.palindrome/main.js
function isPalindromeByReverseAndCompare(str) {
  const rev = str.split("").reverse().join("");
  return str === rev;
}

function isPalindromeByComparingEachLetter(str) {
  return str.split("").every((char, i) => {
    return char === str[str.length - i - 1]; // Not ideal because it could stop in the middle (but it goes until the very last character)
  });
}

function isPalindromeByComparingEachLetterOptimized(str) {
  const len = str.length;
  for (let i = 0; i < len / 2; i++) {
    if (str[i] !== str[len - 1 - i]) return false; // Ideal, because it goes until the middle only
  }
  return true;
}

// Testing
const test = require('node:test');
const assert = require('node:assert');

test('a palindrome', () => {
  assert.strictEqual(isPalindromeByReverseAndCompare('abba'), true);
  assert.strictEqual(isPalindromeByComparingEachLetter('abba'), true);
  assert.strictEqual(isPalindromeByComparingEachLetterOptimized('abba'), true);
});


test('a non-palindrome', () => {
  assert.strictEqual(isPalindromeByReverseAndCompare(' abba'), false);
  assert.strictEqual(isPalindromeByComparingEachLetter(' abba'), false);
  assert.strictEqual(isPalindromeByComparingEachLetterOptimized(' abba'), false);
});

File: algorithm/challenges/02.palindrome/main.clj
(ns main
  (:require [clojure.test :as test]
            [clojure.string :as str]))

(defn palindrome?-with-reverse
  [word]
  (= (str/reverse word) word))

(defn palindrome?-with-map
  [word]
  (->> (map-indexed
        (fn [i char]
          (let [associated-char (get word (- (count word) i 1))]
            (= char associated-char))) ; not optimal because it could go until the middle of the vector only
        word)
       (every? true?)))

(test/deftest palindrome?-test
  (test/testing "a valid palindrome"
    (test/is (true? (palindrome?-with-reverse "abIba")))
    (test/is (true? (palindrome?-with-map "abIba"))))
  (test/testing "an invalid palindrome"
    (test/is (false? (palindrome?-with-reverse " abba")))
    (test/is (false? (palindrome?-with-map " abba")))))

(palindrome?-test)
